ui.py

```diff
@@ -1,75 +1,104 @@
 import textwrap
+from functools import partial
 
 import mdpopups
 import sublime
-from LSP.plugin.core.typing import List, Tuple
+from LSP.plugin.core.typing import List, Optional
 
 from .types import CopilotPayloadCompletion
-from .utils import get_copilot_view_setting, reformat, set_copilot_view_setting
+from .utils import clamp, get_copilot_view_setting, reformat, set_copilot_view_setting
 
 
-class Completion:
+class ViewCompletionManager:
     def __init__(self, view: sublime.View) -> None:
         self.view = view
 
     @property
     def is_visible(self) -> bool:
-        return bool(get_copilot_view_setting(self.view, "is_visible") or False)
+        """Whether Copilot's completion popup is visible."""
+        return get_copilot_view_setting(self.view, "is_visible", False)
 
     @property
-    def region(self) -> Tuple[int, int]:
-        return get_copilot_view_setting(self.view, "region") or (-1, -1)
+    def completions(self) -> List[CopilotPayloadCompletion]:
+        """All `completions` in the view."""
+        return get_copilot_view_setting(self.view, "completions", [])
 
     @property
-    def text(self) -> str:
-        return get_copilot_view_setting(self.view, "text") or ""
+    def completion_index(self) -> int:
+        """The current index of the chosen completion."""
+        return get_copilot_view_setting(self.view, "completion_index", 0)
 
     @property
-    def display_text(self) -> str:
-        return get_copilot_view_setting(self.view, "display_text") or ""
-
-    @property
-    def uuid(self) -> str:
-        return get_copilot_view_setting(self.view, "uuid") or ""
-
-    def get_display_text(self, region: Tuple[int, int], raw_display_text: str) -> str:
-        if "\n" in raw_display_text:
-            return raw_display_text
-
-        if self.view.classify(region[1]) & sublime.CLASS_LINE_END:
-            return raw_display_text
-
-        current_line = self.view.line(region[1])
-        following_text = self.view.substr(sublime.Region(region[0], current_line.end())).strip()
-        index = raw_display_text.find(following_text)
+    def current_completion(self) -> Optional[CopilotPayloadCompletion]:
+        """The current chosen `completion`."""
+        return self.completions[self.completion_index] if self.completions else None
+
+    def show_previous_completion(self) -> None:
+        """Set `completion_index` to be for the previous completion."""
+        self.show(
+            None,
+            self.completion_index - 1,
```

Comment:
Let's use keyword arguments to avoid `None`, e.g. `completion_index=...`

Reply:
Done in [c39d438](https://github.com/TheSecEng/LSP-copilot/pull/13/commits/c39d4389eb499ccfbff25b2792ff9608569f7ed4)